### PR TITLE
[FIXED JENKINS-31768] - Remove monitor lock from setNumberExecutors to avoid deadlock

### PR DESCRIPTION
**Pull request by fengxx**
fix JENKINS-31768 dead lock while removing computer

https://issues.jenkins-ci.org/browse/JENKINS-31768

### CODE REVIEW

**Comment by jtnord** [comment]
this introduces a race condition where 2 users both try to change the number of executors, and the result may be a configuration out of step with reality.

This is due to the `diff` count being calculated outside of the lock where the executor is set (and executors are added or removed).

**Comment by fengxx** [comment]
I added comments about calling path, it is protected by Queue.withLock in most cases.

**Comment by oleg-nenashev** [comment]
<code> addNewExecutorIfNecessary()</code> is not synchronized, so the fix does not seem to be safe . Maybe you could put the second method section into another synchronization block

**Comment by fengxx** [comment]
From the calling path point of view, I think adding new executors are all protected by Queue.withLock

**Comment by olivergondza** [comment]
I wonder if we still need the code that kills idle executors after on demand thread creation introduced in 1.536.

**Comment by olivergondza** [comment]
> I wonder if we still need the code that kills idle executors after on demand thread creation introduced in 1.536.

We do, the executors are not killed eagerly as I assumed.

**Comment by olivergondza** [comment]
Anyway, removing the synchronization is not the correct fix. Few ideas:

- The other thread is blocked in `Nodes#addNode`. The queue lock is commented with:

> TODO we should not need to lock the queue for adding nodes but until we have a way to update the computer list for just the new node

- We can make `Queue._withLock` smarter not to cause deadlocks using `tryLock` and wait approach.

@stephenc WDYT?

**Comment by stephenc** [comment]
Remind me on Friday when I'm back from vacation

**Comment by olivergondza** [comment]
@stephenc, how was your vacation?

**Comment by oleg-nenashev** [comment]
@stephenc How was your another vacation? ^_^

**Comment by stephenc** [comment]
Ok, I've added this to my reminders list... vacation was great BTW

File: core/src/main/java/hudson/model/Computer.java [hunk 1]
```diff
@@ -828,7 +828,19 @@ protected void kill() {
...
+     * Computer.doConfigSubmit -> Computer.replaceBy ->Jenkins.setNodes* ->Computer.setNode
+     * AbstractCIBase.updateComputerList->Computer.inflictMortalWound*
+     * AbstractCIBase.updateComputerList->AbstractCIBase.updateComputer* ->Computer.setNode
+     * AbstractCIBase.updateComputerList->AbstractCIBase.killComputer->Computer.kill
+     * Computer.constructor->Computer.setNode
+     * Computer.kill is called after numExecutors set to zero(Computer.inflictMortalWound) so not need the Queue.lock
+     *
+     * @param number of executors
+     */
+    private void setNumExecutors(int n) {
```
Comment (stephenc) [inline core/src/main/java/hudson/model/Computer.java]: Can we add a `@GuardedBy("hudson.model.Queue.lock")` annotation. Hopefully that way findbugs can confirm your analysis (if it does, we would need to suppress FB warnings on `Computer.kill`) and that way we can protect against further changes breaking the protection

**Comment by oleg-nenashev** [comment]
LGTM 👍

**Comment by oleg-nenashev** [comment]
CC @jenkinsci/code-reviewers I'm about merging it to the next weekly

**Comment by stephenc** [comment]
@oleg-nenashev I've merged this. It has had long enough a chance for others to chime in. It was really blocking on my review and I was happy once the `@GuardedBy` annotation was applied

**Comment by oleg-nenashev** [comment]
@stephenc Well, your merge caused a race condition with the release process :(
Not a big deal, I hope

**Pull request merged**